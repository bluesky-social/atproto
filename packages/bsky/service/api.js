'use strict' /* eslint-disable */

require('dd-trace') // Only works with commonjs
  .init({ logInjection: true })
  .tracer.use('express', {
    hooks: {
      request: (span, req) => {
        maintainXrpcResource(span, req)
      },
    },
  })

// Tracer code above must come before anything else
const path = require('path')
const assert = require('assert')
const { CloudfrontInvalidator } = require('@atproto/aws')
const {
  DatabaseCoordinator,
  PrimaryDatabase,
  ServerConfig,
  BskyAppView,
  ViewMaintainer,
  makeAlgos,
  PeriodicModerationActionReversal,
} = require('@atproto/bsky')

const main = async () => {
  const env = getEnv()
  assert(env.dbPrimaryPostgresUrl, 'missing configuration for db')
  const db = new DatabaseCoordinator({
    schema: env.dbPostgresSchema,
    primary: {
      url: env.dbPrimaryPostgresUrl,
      poolSize: env.dbPrimaryPoolSize || env.dbPoolSize,
      poolMaxUses: env.dbPoolMaxUses,
      poolIdleTimeoutMs: env.dbPoolIdleTimeoutMs,
    },
    replicas: env.dbReplicaPostgresUrls?.map((url, i) => {
      return {
        url,
        poolSize: env.dbPoolSize,
        poolMaxUses: env.dbPoolMaxUses,
        poolIdleTimeoutMs: env.dbPoolIdleTimeoutMs,
        tags: getTagsForIdx(env.dbReplicaTags, i),
      }
    }),
  })
  const cfg = ServerConfig.readEnv({
    port: env.port,
    version: env.version,
    dbPrimaryPostgresUrl: env.dbPrimaryPostgresUrl,
    dbReplicaPostgresUrls: env.dbReplicaPostgresUrls,
    dbReplicaTags: env.dbReplicaTags,
    dbPostgresSchema: env.dbPostgresSchema,
    publicUrl: env.publicUrl,
    didPlcUrl: env.didPlcUrl,
    imgUriSalt: env.imgUriSalt,
    imgUriKey: env.imgUriKey,
    imgUriEndpoint: env.imgUriEndpoint,
    blobCacheLocation: env.blobCacheLocation,
  })
  const cfInvalidator = env.cfDistributionId
    ? new CloudfrontInvalidator({
        distributionId: env.cfDistributionId,
        pathPrefix: cfg.imgUriEndpoint && new URL(cfg.imgUriEndpoint).pathname,
      })
    : undefined
  const algos = env.feedPublisherDid ? makeAlgos(env.feedPublisherDid) : {}
  const bsky = BskyAppView.create({
    db,
    config: cfg,
    imgInvalidator: cfInvalidator,
    algos,
  })
  // separate db needed for more permissions
  const migrateDb = new PrimaryDatabase({
    url: env.dbMigratePostgresUrl,
    schema: env.dbPostgresSchema,
    poolSize: 2,
  })
  const viewMaintainer = new ViewMaintainer(migrateDb)
  const viewMaintainerRunning = viewMaintainer.run()

  const periodicModerationActionReversal = new PeriodicModerationActionReversal(
    bsky.ctx,
  )
  const periodicModerationActionReversalRunning =
    periodicModerationActionReversal.run()

  await bsky.start()
  // Graceful shutdown (see also https://aws.amazon.com/blogs/containers/graceful-shutdowns-with-ecs/)
  process.on('SIGTERM', async () => {
    // Gracefully shutdown periodic-moderation-action-reversal before destroying bsky instance
    periodicModerationActionReversal.destroy()
    await periodicModerationActionReversalRunning
    await bsky.destroy()
    viewMaintainer.destroy()
    await viewMaintainerRunning
    await migrateDb.close()
  })
}

const getEnv = () => ({
  port: parseInt(process.env.PORT),
  version: process.env.BSKY_VERSION,
  dbMigratePostgresUrl:
    process.env.DB_MIGRATE_POSTGRES_URL || process.env.DB_PRIMARY_POSTGRES_URL,
  dbPrimaryPostgresUrl: process.env.DB_PRIMARY_POSTGRES_URL,
  dbPrimaryPoolSize: maybeParseInt(process.env.DB_PRIMARY_POOL_SIZE),
  dbReplicaPostgresUrls: process.env.DB_REPLICA_POSTGRES_URLS
    ? process.env.DB_REPLICA_POSTGRES_URLS.split(',')
    : undefined,
  dbReplicaTags: {
    '*': getTagIdxs(process.env.DB_REPLICA_TAGS_ANY), // e.g. DB_REPLICA_TAGS_ANY=0,1
    timeline: getTagIdxs(process.env.DB_REPLICA_TAGS_TIMELINE),
    feed: getTagIdxs(process.env.DB_REPLICA_TAGS_FEED),
    search: getTagIdxs(process.env.DB_REPLICA_TAGS_SEARCH),
    thread: getTagIdxs(process.env.DB_REPLICA_TAGS_THREAD),
  },
  dbPostgresSchema: process.env.DB_POSTGRES_SCHEMA || undefined,
  dbPoolSize: maybeParseInt(process.env.DB_POOL_SIZE),
  dbPoolMaxUses: maybeParseInt(process.env.DB_POOL_MAX_USES),
  dbPoolIdleTimeoutMs: maybeParseInt(process.env.DB_POOL_IDLE_TIMEOUT_MS),
  publicUrl: process.env.PUBLIC_URL,
  didPlcUrl: process.env.DID_PLC_URL,
  imgUriSalt: process.env.IMG_URI_SALT,
  imgUriKey: process.env.IMG_URI_KEY,
  imgUriEndpoint: process.env.IMG_URI_ENDPOINT,
  blobCacheLocation: process.env.BLOB_CACHE_LOC,
  cfDistributionId: process.env.CF_DISTRIBUTION_ID,
  feedPublisherDid: process.env.FEED_PUBLISHER_DID,
})

/**
 * @param {Record<string, number[]>} tags
 * @param {number} idx
 */
const getTagsForIdx = (tagMap, idx) => {
  const tags = []
  for (const [tag, indexes] of Object.entries(tagMap)) {
    if (indexes.includes(idx)) {
      tags.push(tag)
    }
  }
  return tags
}

/**
 * @param {string} str
 */
const getTagIdxs = (str) => {
  return str ? str.split(',').map((item) => parseInt(item, 10)) : []
}

const maybeParseInt = (str) => {
  const parsed = parseInt(str)
  return isNaN(parsed) ? undefined : parsed
}

const maintainXrpcResource = (span, req) => {
  // Show actual xrpc method as resource rather than the route pattern
  if (span && req.originalUrl?.startsWith('/xrpc/')) {
    span.setTag(
      'resource.name',
      [
        req.method,
        path.posix.join(req.baseUrl || '', req.path || '', '/').slice(0, -1), // Ensures no trailing slash
      ]
        .filter(Boolean)
        .join(' '),
    )
  }
}

main()
