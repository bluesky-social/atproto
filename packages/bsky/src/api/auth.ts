import express from 'express'
import * as uint8arrays from 'uint8arrays'
import { AuthRequiredError } from '@atproto/xrpc-server'

const BASIC = 'Basic '
const BEARER = 'Bearer '

// @TODO(bsky) treating did as a bearer, just a placeholder for now.
export const authVerifier = (ctx: {
  req: express.Request
  res: express.Response
}) => {
  const { authorization = '' } = ctx.req.headers
  if (!authorization.startsWith(BEARER)) {
    throw new AuthRequiredError()
  }
  const did = authorization.replace(BEARER, '').trim()
  if (!did.startsWith('did:')) {
    throw new AuthRequiredError()
  }
  return { credentials: { did } }
}

export const authOptionalVerifier = (ctx: {
  req: express.Request
  res: express.Response
}) => {
  if (!ctx.req.headers.authorization) {
    return { credentials: { did: null } }
  }
  return authVerifier(ctx)
}

export const adminVerifier =
  (adminPassword: string) =>
  (ctx: { req: express.Request; res: express.Response }) => {
    const { authorization = '' } = ctx.req.headers
    const parsed = parseBasicAuth(authorization)
    if (!parsed) {
      throw new AuthRequiredError()
    }
    const { username, password } = parsed
    if (username !== 'admin' || password !== adminPassword) {
      throw new AuthRequiredError()
    }
    return { credentials: { admin: true } }
  }

export const parseBasicAuth = (
  token: string,
): { username: string; password: string } | null => {
  if (!token.startsWith(BASIC)) return null
  const b64 = token.slice(BASIC.length)
  let parsed: string[]
  try {
    parsed = uint8arrays
      .toString(uint8arrays.fromString(b64, 'base64pad'), 'utf8')
      .split(':')
  } catch (err) {
    return null
  }
  const [username, password] = parsed
  if (!username || !password) return null
  return { username, password }
}
