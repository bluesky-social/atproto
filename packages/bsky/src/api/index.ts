import { Server } from '../lexicon'
import AppContext from '../context'
import getTimeline from './app/bsky/feed/getTimeline'
import getAuthorFeed from './app/bsky/feed/getAuthorFeed'
import getLikes from './app/bsky/feed/getLikes'
import getPostThread from './app/bsky/feed/getPostThread'
import getProfile from './app/bsky/actor/getProfile'
import getProfiles from './app/bsky/actor/getProfiles'
import getRepostedBy from './app/bsky/feed/getRepostedBy'
import getFollowers from './app/bsky/graph/getFollowers'
import getFollows from './app/bsky/graph/getFollows'
import searchActors from './app/bsky/actor/searchActors'
import searchActorsTypeahead from './app/bsky/actor/searchActorsTypeahead'
import getSuggestions from './app/bsky/actor/getSuggestions'
import unspecced from './app/bsky/unspecced'

export * as health from './health'

export * as blobResolver from './blob-resolver'

export default function (server: Server, ctx: AppContext) {
  getTimeline(server, ctx)
  getAuthorFeed(server, ctx)
  getLikes(server, ctx)
  getPostThread(server, ctx)
  getProfile(server, ctx)
  getProfiles(server, ctx)
  getRepostedBy(server, ctx)
  getFollowers(server, ctx)
  getFollows(server, ctx)
  searchActors(server, ctx)
  searchActorsTypeahead(server, ctx)
  getSuggestions(server, ctx)
  unspecced(server, ctx)
  return server
}
