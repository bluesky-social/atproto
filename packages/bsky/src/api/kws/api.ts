import express, { RequestHandler } from 'express'
import { httpLogger as log } from '../../logger'
import {
  AppContextWithKwsClient,
  KwsVerificationIntermediateQuery,
  KwsVerificationQuery,
  verificationIntermediateQuerySchema,
} from './types'
import {
  createStashEvent,
  getClientUa,
  parseExternalPayload,
  parseStatus,
  validateSignature,
} from './util'

const validateRequest = (
  ctx: AppContextWithKwsClient,
  req: express.Request,
): KwsVerificationQuery => {
  try {
    const intermediate: KwsVerificationIntermediateQuery =
      verificationIntermediateQuerySchema.parse({
        externalPayload: req.query.externalPayload,
        signature: req.query.signature,
        status: req.query.status,
      })

    const data = `${intermediate.status}:${intermediate.externalPayload}`
    validateSignature(
      ctx.cfg.kws.verificationSecret,
      data,
      intermediate.signature,
    )

    return {
      ...intermediate,
      externalPayload: parseExternalPayload(intermediate.externalPayload),
      status: parseStatus(intermediate.status),
    }
  } catch (err) {
    throw new Error('Invalid KWS API request', { cause: err })
  }
}

export const verificationHandler =
  (ctx: AppContextWithKwsClient): RequestHandler =>
  async (req: express.Request, res: express.Response) => {
    let actorDid: string | undefined
    try {
      const query = validateRequest(ctx, req)
      const {
        externalPayload,
        status: { verified },
      } = query
      if (!verified) {
        throw new Error(
          'Unexpected KWS verification response call with unverified status',
        )
      }

      const { actorDid: externalPayloadActorDid, attemptId } = externalPayload
      actorDid = externalPayloadActorDid
      // Assumes `app.set('trust proxy', ...)` configured with `true` or specific values.
      const completeIp = req.ip
      const completeUa = getClientUa(req)
      await createStashEvent(ctx, {
        actorDid,
        attemptId,
        completeIp,
        completeUa,
        status: 'assured',
      })
      return res
        .status(302)
        .setHeader(
          'Location',
          `${ctx.cfg.kws.redirectUrl}?${new URLSearchParams({ actorDid, result: 'success' })}`,
        )
        .end()
    } catch (err) {
      log.error({ err }, 'Failed to handle KWS verification response')

      return res
        .status(302)
        .setHeader(
          'Location',
          `${ctx.cfg.kws.redirectUrl}?${new URLSearchParams({ ...(actorDid ? { actorDid } : {}), result: 'unknown' })}`,
        )
        .end()
    }
  }
