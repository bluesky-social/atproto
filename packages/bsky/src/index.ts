import express from 'express'
import http from 'http'
import { AddressInfo } from 'net'
import events from 'events'
import { createHttpTerminator, HttpTerminator } from 'http-terminator'
import cors from 'cors'
import { DidResolver } from '@atproto/did-resolver'
import API, { health, blobResolver } from './api'
import Database from './db'
import * as error from './error'
import { loggerMiddleware } from './logger'
import { ServerConfig } from './config'
import { createServer } from './lexicon'
import { ImageUriBuilder } from './image/uri'
import { BlobDiskCache, ImageProcessingServer } from './image/server'
import { createServices } from './services'
import AppContext from './context'
import { RepoSubscription } from './subscription/repo'

export type { ServerConfigValues } from './config'
export { ServerConfig } from './config'
export { Database } from './db'
export { AppContext } from './context'

export class BskyAppView {
  public ctx: AppContext
  public app: express.Application
  public sub?: RepoSubscription
  public server?: http.Server
  private terminator?: HttpTerminator

  constructor(opts: {
    ctx: AppContext
    app: express.Application
    sub?: RepoSubscription
  }) {
    this.ctx = opts.ctx
    this.app = opts.app
    this.sub = opts.sub
  }

  static create(opts: { db: Database; config: ServerConfig }): BskyAppView {
    const { db, config } = opts
    const app = express()
    app.use(cors())
    app.use(loggerMiddleware)

    const didResolver = new DidResolver({ plcUrl: config.didPlcUrl })

    const imgUriBuilder = new ImageUriBuilder(
      config.imgUriEndpoint || `${config.publicUrl}/image`,
      config.imgUriSalt,
      config.imgUriKey,
    )

    const services = createServices({
      imgUriBuilder,
      didResolver,
    })

    const ctx = new AppContext({
      db,
      cfg: config,
      services,
      imgUriBuilder,
      didResolver,
    })

    let imgProcessingServer: ImageProcessingServer | undefined
    if (!config.imgUriEndpoint) {
      const imgProcessingCache = new BlobDiskCache(config.blobCacheLocation)
      imgProcessingServer = new ImageProcessingServer(ctx, imgProcessingCache)
    }

    let server = createServer({
      validateResponse: config.debugMode,
      payload: {
        jsonLimit: 100 * 1024, // 100kb
        textLimit: 100 * 1024, // 100kb
        blobLimit: 5 * 1024 * 1024, // 5mb
      },
    })

    server = API(server, ctx)

    app.use(health.createRouter(ctx))
    app.use(blobResolver.createRouter(ctx))
    if (imgProcessingServer) {
      app.use('/image', imgProcessingServer.app)
    }
    app.use(server.xrpc.router)
    app.use(error.handler)

    const sub = config.repoProvider
      ? new RepoSubscription(ctx, config.repoProvider, config.repoSubLockId)
      : undefined

    return new BskyAppView({ ctx, app, sub })
  }

  async start(): Promise<http.Server> {
    const server = this.app.listen(this.ctx.cfg.port)
    this.server = server
    this.terminator = createHttpTerminator({ server })
    await events.once(server, 'listening')
    const { port } = server.address() as AddressInfo
    this.ctx.cfg.assignPort(port)
    this.sub?.run() // Don't await, backgrounded
    return server
  }

  async destroy(): Promise<void> {
    await this.sub?.destroy()
    await this.terminator?.terminate()
    await this.ctx.db.close()
  }
}

export default BskyAppView
