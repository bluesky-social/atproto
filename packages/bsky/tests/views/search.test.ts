import { AtpAgent } from '@atproto/api'
import { SeedClient, TestNetwork, basicSeed } from '@atproto/dev-env'
import { DatabaseSchema } from '../../src'
import { ids } from '../../src/lexicon/lexicons'
import { forSnapshot } from '../_util'

const TAG_HIDE = 'hide'

describe('appview search', () => {
  let network: TestNetwork
  let agent: AtpAgent
  let sc: SeedClient
  let post0: Awaited<ReturnType<SeedClient['post']>>
  let post1: Awaited<ReturnType<SeedClient['post']>>
  let post2: Awaited<ReturnType<SeedClient['post']>>
  let posts: Awaited<ReturnType<SeedClient['post']>>[]

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'bsky_views_search',
      bsky: {
        searchTagsHide: new Set([TAG_HIDE]),
      },
    })
    agent = network.bsky.getClient()
    sc = network.getSeedClient()
    await basicSeed(sc)

    post0 = await sc.post(sc.dids.alice, 'good doggo')
    post1 = await sc.post(sc.dids.alice, 'bad doggo')
    post2 = await sc.post(sc.dids.alice, 'cute doggo')
    posts = [post0, post1, post2]
    await network.processAll()
  })

  afterAll(async () => {
    await network.close()
  })

  it('finds posts by search term and top sort', async () => {
    const res = await agent.app.bsky.feed.searchPosts(
      { q: 'doggo', sort: 'top' },
      {
        headers: await network.serviceHeaders(
          sc.dids.alice,
          ids.AppBskyFeedSearchPosts,
        ),
      },
    )

    expect(res.data.posts.length).toBe(posts.length)
    expect(forSnapshot(res.data.posts)).toMatchSnapshot()
  })

  it('finds posts by search term and latest sort', async () => {
    const res = await agent.app.bsky.feed.searchPosts(
      { q: 'doggo', sort: 'latest' },
      {
        headers: await network.serviceHeaders(
          sc.dids.alice,
          ids.AppBskyFeedSearchPosts,
        ),
      },
    )

    expect(res.data.posts.length).toBe(posts.length)
    expect(forSnapshot(res.data.posts)).toMatchSnapshot()
  })

  describe('hiding results', () => {
    const visiblePosts = [post0, post2]

    beforeAll(async () => {
      await createTag(network.bsky.db.db, {
        uri: post1.ref.uriStr,
        val: TAG_HIDE,
      })
      await network.processAll()
    })

    it('finds posts by search term and top sort', async () => {
      const res = await agent.app.bsky.feed.searchPosts(
        { q: 'doggo', sort: 'top' },
        {
          headers: await network.serviceHeaders(
            sc.dids.alice,
            ids.AppBskyFeedSearchPosts,
          ),
        },
      )

      expect(res.data.posts.length).toBe(visiblePosts.length)
      expect(forSnapshot(res.data.posts)).toMatchSnapshot()
    })

    it('finds posts by search term and latest sort', async () => {
      const res = await agent.app.bsky.feed.searchPosts(
        { q: 'doggo', sort: 'latest' },
        {
          headers: await network.serviceHeaders(
            sc.dids.alice,
            ids.AppBskyFeedSearchPosts,
          ),
        },
      )

      expect(res.data.posts.length).toBe(visiblePosts.length)
      expect(forSnapshot(res.data.posts)).toMatchSnapshot()
    })
  })
})

const createTag = async (
  db: DatabaseSchema,
  opts: {
    uri: string
    val: string
  },
) => {
  await db
    .updateTable('record')
    .set({
      tags: JSON.stringify([opts.val]),
    })
    .where('uri', '=', opts.uri)
    .returningAll()
    .execute()
}
