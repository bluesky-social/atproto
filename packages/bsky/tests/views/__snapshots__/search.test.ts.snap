// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appview search finds posts by search term and latest sort 1`] = `
Array [
  Object {
    "author": Object {
      "associated": Object {
        "activitySubscription": Object {
          "allowSubscriptions": "followers",
        },
      },
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "bookmarkCount": 0,
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "quoteCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "cute doggo",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(0)",
    "viewer": Object {
      "bookmarked": false,
      "embeddingDisabled": false,
      "threadMuted": false,
    },
  },
  Object {
    "author": Object {
      "associated": Object {
        "activitySubscription": Object {
          "allowSubscriptions": "followers",
        },
      },
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "bookmarkCount": 0,
    "cid": "cids(3)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "quoteCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "bad doggo",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(2)",
    "viewer": Object {
      "bookmarked": false,
      "embeddingDisabled": false,
      "threadMuted": false,
    },
  },
  Object {
    "author": Object {
      "associated": Object {
        "activitySubscription": Object {
          "allowSubscriptions": "followers",
        },
      },
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "bookmarkCount": 0,
    "cid": "cids(4)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "quoteCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "good doggo",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(3)",
    "viewer": Object {
      "bookmarked": false,
      "embeddingDisabled": false,
      "threadMuted": false,
    },
  },
]
`;

exports[`appview search finds posts by search term and top sort 1`] = `
Array [
  Object {
    "author": Object {
      "associated": Object {
        "activitySubscription": Object {
          "allowSubscriptions": "followers",
        },
      },
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "bookmarkCount": 0,
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "quoteCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "cute doggo",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(0)",
    "viewer": Object {
      "bookmarked": false,
      "embeddingDisabled": false,
      "threadMuted": false,
    },
  },
  Object {
    "author": Object {
      "associated": Object {
        "activitySubscription": Object {
          "allowSubscriptions": "followers",
        },
      },
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "bookmarkCount": 0,
    "cid": "cids(3)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "quoteCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "bad doggo",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(2)",
    "viewer": Object {
      "bookmarked": false,
      "embeddingDisabled": false,
      "threadMuted": false,
    },
  },
  Object {
    "author": Object {
      "associated": Object {
        "activitySubscription": Object {
          "allowSubscriptions": "followers",
        },
      },
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "bookmarkCount": 0,
    "cid": "cids(4)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "quoteCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "good doggo",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(3)",
    "viewer": Object {
      "bookmarked": false,
      "embeddingDisabled": false,
      "threadMuted": false,
    },
  },
]
`;

exports[`appview search hiding results finds posts by search term and latest sort 1`] = `
Array [
  Object {
    "author": Object {
      "associated": Object {
        "activitySubscription": Object {
          "allowSubscriptions": "followers",
        },
      },
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "bookmarkCount": 0,
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "quoteCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "cute doggo",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(0)",
    "viewer": Object {
      "bookmarked": false,
      "embeddingDisabled": false,
      "threadMuted": false,
    },
  },
  Object {
    "author": Object {
      "associated": Object {
        "activitySubscription": Object {
          "allowSubscriptions": "followers",
        },
      },
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "bookmarkCount": 0,
    "cid": "cids(3)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "quoteCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "good doggo",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(2)",
    "viewer": Object {
      "bookmarked": false,
      "embeddingDisabled": false,
      "threadMuted": false,
    },
  },
]
`;

exports[`appview search hiding results finds posts by search term and top sort 1`] = `
Array [
  Object {
    "author": Object {
      "associated": Object {
        "activitySubscription": Object {
          "allowSubscriptions": "followers",
        },
      },
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "bookmarkCount": 0,
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "quoteCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "cute doggo",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(0)",
    "viewer": Object {
      "bookmarked": false,
      "embeddingDisabled": false,
      "threadMuted": false,
    },
  },
  Object {
    "author": Object {
      "associated": Object {
        "activitySubscription": Object {
          "allowSubscriptions": "followers",
        },
      },
      "avatar": "https://bsky.public.url/img/avatar/plain/user(1)/cids(1)@jpeg",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "src": "user(0)",
          "uri": "record(1)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "muted": false,
      },
    },
    "bookmarkCount": 0,
    "cid": "cids(3)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 0,
    "quoteCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "good doggo",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(2)",
    "viewer": Object {
      "bookmarked": false,
      "embeddingDisabled": false,
      "threadMuted": false,
    },
  },
]
`;
